### ISSUE TITLE
Bluetooth audio not working for call audio after latest update

### ISSUE THREAD

**Issue by marissarousseau** [GrapheneOS/os_issue_tracker#217]
I upgraded to the latest release of GrapheneOS on the Pixel 3 just a few hours ago. Since installing the update, my Bluetooth does not work for any call audio, and I am unable to hear anything coming through my bluetooth earbuds.

**Comment by thestinger** [comment]
cc @valldrac

https://github.com/GrapheneOS/os_issue_tracker/issues/137

**Comment by thestinger** [comment]
@marissarousseau Can you try rebooting and trying it a few times. Maybe there's another case where it only has a 1/3 chance of working and this made things worse for you by not aligning the order in both places. Not sure.

**Comment by marissarousseau** [comment]
> @marissarousseau Can you try rebooting and trying it a few times. Maybe there's another case where it only has a 1/3 chance of working and this made things worse for you by not aligning the order in both places. Not sure.

Looks like another reboot did the trick. I tried rebooting before opening the issue but it didn't work, so it might be random chance that it's fixed now.

**Comment by thestinger** [comment]
It's possible there's only a 1/3 chance for it to fail. It would be helpful if you could reboot a few more times and also keep an eye on this to see if it breaks again.

**Comment by thestinger** [comment]
Maybe it just needed to force a reconfiguration after we fixed the order, or maybe it's broken in some other way. Before the fix, the order was random due to an upstream bug relying on malloc addresses. Since hardened_malloc makes that random, the order is always random. There are 3 Bluetooth types in the array so there are a lot of different possible sort orders. It's possible we only partially got the correct order, possible there's something else doing something similar, etc. Lot of things that could still be wrong.

**Comment by marissarousseau** [comment]
> It's possible there's only a 1/3 chance for it to fail. It would be helpful if you could reboot a few more times and also keep an eye on this to see if it breaks again.

Sure, here's what I've seen so far.

First reboot after installing update: Not working
Second: Working
Third: Working
Fourth: Working
Fifth: Working

It looks like it was just the initial update after installing.

**Comment by thestinger** [comment]
I guess it's possible it might have started working in the initial reboot at least if it was repaired or something like that. Seems like it does work properly and just needed some kind of initial push to start working after the change. Not really sure. There might be a real problem remaining though. Probably can't do anything to ease the transition at this point if there's not an issue beyond it adjusting to the new order.

**Comment by 4r7urh** [comment]
first thing: BT audio for music works much better now, this is great. In the car I have spotify playing, with a second app sounding warning messages for road hazards. When a road hazard comes up, now spotify volume is decreased, warning message is played and afterwards spotify volume returns to normal. This did not work previously, there music partially was breaking up and volume stayed decreased.

- Issue with call audio is not fixed by rebooting on my device (pixel 3a). Tried that at least 5 times.
- After a reboot, if you immediately make a phone call, call audio can be heard in the headset. If you afterwards wait for some minutes, call audio cannot be heard in the next call, this stays like this going forward.
- Deleting paired headsets and reconnecting them again did not show any improvement.
- Also tried to toggle parameter for BT hardware offload, no improvements were noticed.
- Tried several different headsets and one car, on all devices same problem with call audio exists.

**Comment by thestinger** [comment]
Did call audio work before the update?

**Comment by thestinger** [comment]
Also please provide the information @valldrac suggested in the other thread:

https://github.com/GrapheneOS/os_issue_tracker/issues/137#issuecomment-633990360

**Comment by 4r7urh** [comment]
yes. I bought my phone end of March and immediately installed GrapheneOS, call audio worked fine between then and yesterday. I've been using BT headsets quite heavily, in the car and at my desk.

Settings > Sound > Play-media-to 
I don't find this option on my phone, as it runs in German language. Audio settings, where I change e.g. the volume has no setting option that comes close to this. Will run the adb command tomorrow, has been a long day in my timezone.

As mentioned above, when a headset is connected, music is played, call audio not

**Comment by ChickenSupreme** [comment]
Just to add another data point, I had the same issue as marissarousseau on my Pixel 3a - i.e. bluetooth call audio stopped working after the update. I rebooted as advised and it's working fine now. Note that bluetooth music was working both before and after the update.

**Comment by thestinger** [comment]
@4r7urh can you try a few more reboots?

**Comment by 4r7urh** [comment]
did 5 more reboots and fetched the audio policy:
[audio_policy 1.txt](https://github.com/GrapheneOS/os_issue_tracker/files/4687070/audio_policy.1.txt)

Did one more reboot. BT call audio worked after the reboot and on the next try about 20 minutes later. 
After the call I paired another BT headset, this immediately caused call audio to stop working, regardless which headset was connected to the phone. This is the audio policy after the last pairing:
[audio_policy 2.txt](https://github.com/GrapheneOS/os_issue_tracker/files/4687104/audio_policy.2.txt)

Hope this helps

**Comment by valldrac** [comment]
Thank you so much @4r7urh. There's a bug in the Bluetooth routing for voice calls. So basically we fixed the audio streaming and break the calls. Working on the fix.

**Comment by ahowell8** [comment]
Didn't see this mentioned, but received a call when the BT on and not working (didn't get a log) and the audio ringtone had a ton of static and choppy sounding - like digital audio from a very low bandwidth stream. I picked up the phone and BT showed to still be connected, voice was fine and coming through phone speakers.

**Comment by JeffAlyanak** [comment]
I can confirm the same behaviour as @marissarousseau, the first reboot after the update on my Pixel 3A broke bluetooth call audio, on subsequent reboots the issue was never present again.

**Comment by hsanjuan** [comment]
I have this problem on the Pixel 3. Reboot fixes bluetooth it for the first call. But it later breaks for other calls (not sure when?) until the next reboot. Same with and without HW offloading. Unlike calls, media playing seems to work fine.

At least I observed once that a bluetooth worked on a call (right after a reboot). Then turned Airplane mode on, then off, and then bluetooth did not work in the next call.

For my this is right now a pretty bad regression, as I did not have any bluetooth audio problems before (with HW offloading disabled).

**Comment by marissarousseau** [comment]
Just following up with my original issue. I have rebooted a few more times since the issue fixed itself earlier, everything is working fine for me still. No issues with media/call audio on Bluetooth.

**Comment by sempervictus** [comment]
I'm on the 4th reboot, and BT works for everything _except_ phone calls themselves. Even dial-tones for pressing the keys come through bluetooth, but all call audio, device only. Reboots, re-pairings, and ritual sacrifice have no effect so far. This actually makes the phone illegal to use in many states while driving, careful if you're experiencing this and live in one of those states like i do.

**Comment by cyrinux** [comment]
Hi, on a pixel 3a, I got same as @sempervictus , bluetooth for audio works for all but not for call. Phone app say bluetooth is enable but sound out from normal speaker. Even after disconnect/connect and re-paring.

**Comment by sempervictus** [comment]
I'm on the pixel3XL by the way.
@thestinger: any debug data we can pull to help determine whats afoot and how to get it back to working state?

**Comment by thestinger** [comment]
I need help working on it. The attempted fix was from @valldrac and appears to have partially resolved the issue while somehow causing other issues. I don't really understand how his fix caused these other issues. It doesn't make sense to me based on the change. I'm not particularly inclined to revert the fix. I need help making it into a proper fix rather than just reverting it and going back to having the other problems instead.

**Comment by thestinger** [comment]
I don't really have time to work on all these things myself. Others need to help with the development work. It's great that @valldrac figured out part of what was wrong, but you can see that there are consequences to barely having any development resources and not having a group of people developing / testing things. He fixed the problems he could verify himself, and it somehow caused new kinds of problems for others. I think he did great work and I don't want to revert the change considering that it DOES make things better for many people.

**Comment by thestinger** [comment]
I know how to get it back to a working state for you: revert the change that fixes issues for others. Don't want to do that. I need help figuring out a solution that works everywhere without side effects. You can see the info @valldrac requested above.

**Comment by sempervictus** [comment]
@thestinger: peeked @ 2a6085d5f0d and not enough context around it to grok whats going on yet. I'll try to read up on where that function is even called and what its returns could mean. I however cannot test builds on my own phone without wiping a production device to use builds signed with my key. If i can come up with a cogent answer about whats happening and how we might fix it, is the process to push a PR to be built with your key in the beta channel, or how does collaborative testing work with limited hardware resources?

**Comment by thestinger** [comment]
Going to revert this change and attempt another workaround for the problem.

**Comment by thestinger** [comment]
https://github.com/GrapheneOS/hardened_malloc/commit/c525803921e622cd50e7556911f339e76c00bff9

Just need to know the executable(s) that are involved and we'll just temporarily disable SLOT_RANDOMIZE for them. This class is apparently removed in Android master anyway probably since it's a broken pile of garbage so a temporary workaround seems like a genuinely good approach to let it solve itself.

**Comment by valldrac** [comment]
The executable that uses libaudiopolicyservice.so is `/system/bin/audioserver`. I reverted my fix and applied your patch, and I cannot reproduce the audio issues anymore; sounds good to me :)

**Comment by sempervictus** [comment]
For edification purposes only: hasn't that thing actually been the subject of attacks/exploits? Whats the reduction in security posture associated with dropping SLOT_RANDOMIZE for a potential target binary which may be reachable by unsanitized external inputs (audio streams/files/etc)?

**Comment by thestinger** [comment]
It's not a particularly important security feature. Also, keep in mind that all these processes have been heavily split up into different sandboxed components, audioserver isn't where files are parsed / decoded.

**Comment by thestinger** [comment]
https://android-developers.googleblog.com/2019/05/queue-hardening-enhancements.html
